src/Microsoft.AspNetCore.Mvc.Razor.Host/Internal/ViewComponentTagHelperChunkVisitor.cs

```diff
@@ -0,0 +1,228 @@
+    {
+        private GeneratedViewComponentTagHelperContext _context;
+        private HashSet<string> _writtenViewComponents;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="ViewComponentTagHelperChunkVisitor"/>. 
+        /// </summary>
+        /// <param name="writer">The <see cref="CSharpCodeWriter"/> the code is generated to.</param>
+        /// <param name="context">The <see cref="CodeGeneratorContext"/> needed to visit chunks.</param>
+        public ViewComponentTagHelperChunkVisitor(CSharpCodeWriter writer, CodeGeneratorContext context) :
+            base(writer, context)
+        {
+            if (writer == null)
+            {
+                throw new ArgumentNullException(nameof(writer));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            _context = new GeneratedViewComponentTagHelperContext();
+            _writtenViewComponents = new HashSet<string>();
+        }
+
+        /// <summary>
+        /// Generates code if the selected <see cref="Chunk"/>, if the chunk is a <see cref="TagHelperChunk"/>
+        /// that describes a view component. If the chunk is a <see cref="ParentChunk"/>, accepts the chunk's children. 
+        /// </summary>
+        /// <param name="chunk">A <see cref="Chunk"/> to accept.</param> 
+        public override void Accept(Chunk chunk)
+        {
+            if (chunk == null)
+            {
+                throw new ArgumentNullException(nameof(chunk));
+            }
+
+            var parentChunk = chunk as ParentChunk;
+            var tagHelperChunk = chunk as TagHelperChunk;
+
+            if (parentChunk != null && !(parentChunk is TagHelperChunk))
+            {
+                Accept(parentChunk.Children);
+            }
+
+            else if (tagHelperChunk != null) 
+            {
+                var viewComponentDescriptors = tagHelperChunk.Descriptors.Where(
+                    descriptor => ViewComponentTagHelperDescriptorConventions.IsViewComponentDescriptor(descriptor));
+                foreach (var descriptor in viewComponentDescriptors)
+                {
+                    base.Accept(chunk);
+                    return;
+                }
+            }
+        }
+
+        /// <summary>
+        /// We generate classes for each view component <see cref="TagHelperDescriptor"/> in the chunk.
+        /// </summary>
+        /// <param name="chunk">The <see cref="TagHelperChunk"/> to visit.</param>
+        protected override void Visit(TagHelperChunk chunk)
+        {
+            var viewComponentDescriptors = chunk.Descriptors.Where(
+                descriptor => ViewComponentTagHelperDescriptorConventions.IsViewComponentDescriptor(descriptor));
+
+            foreach (var descriptor in viewComponentDescriptors)
+            {
+                var descriptorName =
```

Comment:
Instead of differentiating the written viewcomponents by the ViewComponent's name do it by the descriptors type name. In the end, that's what will cause a compilation issue.